XMonad/Util/NamedScratchpad.hs

```diff
@@ -115,30 +118,48 @@ findByName c s = listToMaybe $ filter ((s==) . name) c
 runApplication :: NamedScratchpad -> X ()
 runApplication = spawn . cmd
 
+-- | Runs application which should appear in a specified scratchpad on the workspace it was launched on
+runApplicationHere :: NamedScratchpad -> X ()
+runApplicationHere = spawnHere . cmd
+
 -- | Action to pop up specified named scratchpad
 namedScratchpadAction :: NamedScratchpads -- ^ Named scratchpads configuration
                       -> String           -- ^ Scratchpad name
                       -> X ()
-namedScratchpadAction = someNamedScratchpadAction (\f ws -> f $ head ws)
+namedScratchpadAction = customRunNamedScratchpadAction runApplication
+
+-- | Action to pop up specified named scratchpad, initially starting it on the current workspace.
+spawnHereNamedScratchpadAction :: NamedScratchpads           -- ^ Named scratchpads configuration
+                               -> String                     -- ^ Scratchpad name
+                               -> X ()
+spawnHereNamedScratchpadAction = customRunNamedScratchpadAction runApplicationHere
+
+-- | Action to pop up specified named scratchpad, given a custom way to initially start the application.
+customRunNamedScratchpadAction :: (NamedScratchpad -> X ())  -- ^ Function initially running the application, given the configured @scratchpad@ cmd
+                               -> NamedScratchpads           -- ^ Named scratchpads configuration
+                               -> String                     -- ^ Scratchpad name
+                               -> X ()
+customRunNamedScratchpadAction = someNamedScratchpadAction (\f ws -> f $ head ws)
```

Comment:
How do we know this use of `head` is safe?  Can this code be rewritten without using it, in a way that is more "obviously" safe?